### PR TITLE
Chunked XContent encoding for Metadata

### PR DESCRIPTION
**Pull request by original-brownbear**
Mostly straight-forward chunked encoding implementation for Metadata.

relates #89838

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 1]
```diff
@@ -1318,9 +1322,61 @@ public static Metadata fromXContent(XContentParser parser) throws IOException {
...
+            customs().entrySet()
+                .stream()
+                .filter(cursor -> cursor.getValue().context().contains(context))
+                .map(
+                    cursor -> (ChunkedToXContent) params -> ChunkedToXContentHelper.wrapWithObject(
+                        cursor.getKey(),
+                        cursor.getValue().toXContentChunked(params)
+                    )
+                )
+                .flatMap(s -> StreamSupport.stream(Spliterators.spliteratorUnknownSize(s.toXContentChunked(p), Spliterator.ORDERED), false))
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1375]: I did come up with a flatmap implementation on Iterators once if that helps: https://github.com/elastic/elasticsearch/compare/main...DaveCTurner:elasticsearch:2022-12-06-Iterator-flatmap

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 2]
```diff
@@ -1318,9 +1322,61 @@ public static Metadata fromXContent(XContentParser parser) throws IOException {
...
+        final Iterator<? extends ToXContent> start = context == XContentContext.API
+            ? ChunkedToXContentHelper.startObject("metadata")
+            : Iterators.single((builder, params) -> builder.startObject("meta-data").field("version", version()));
+
+        final Iterator<? extends ToXContent> persSettings = context != XContentContext.API && persistentSettings().isEmpty() == false
+            ? Iterators.single(
+                (builder, params) -> persistentSettings().toXContent(
+                    builder.startObject("settings"),
+                    new ToXContent.MapParams(Collections.singletonMap("flat_settings", "true"))
+                ).endObject()
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1336]: Hmm I'd rather introduce a block and make 3 calls to the builder, this is a bit convoluted to read.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 3]
```diff
@@ -1318,9 +1322,61 @@ public static Metadata fromXContent(XContentParser parser) throws IOException {
...
-    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
-        Builder.toXContent(this, builder, params);
-        return builder;
+    public Iterator<? extends ToXContent> toXContentChunked(ToXContent.Params p) {
+        XContentContext context = XContentContext.valueOf(p.param(CONTEXT_MODE_PARAM, CONTEXT_MODE_API));
+        final Iterator<? extends ToXContent> start = context == XContentContext.API
+            ? ChunkedToXContentHelper.startObject("metadata")
+            : Iterators.single((builder, params) -> builder.startObject("meta-data").field("version", version()));
+
+        final Iterator<? extends ToXContent> persSettings = context != XContentContext.API && persistentSettings().isEmpty() == false
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1331]: Naming nit, let's use the full word
```suggestion
        final Iterator<? extends ToXContent> persistentSettings = context != XContentContext.API && persistentSettings().isEmpty() == false
```

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I left a couple of comments/suggestions

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1336]: ++ fair point :)

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1374]: Yea something like that seems right, but urgh ... I was hoping to find a short and sweet solution. The iterator magic gets a little out of hand at this point :D

**Review by DaveCTurner — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1374]: Do you mean the usage or the implementation? This implementation is nothing compared to the inner workings of `Stream#flatMap`.

**Comment by original-brownbear** [comment]
Thanks David!

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1374]: Implementation and you're definitely right that it's much simpler than what the stream does :)

**Pull request merged**